lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/impl/ThreadPoolStage.java

```diff
@@ -206,14 +208,29 @@ public void run() {
    * Closes resources.
    */
   @Override
-  public void close() throws Exception {
+  public void close() {
+
     if (closed.compareAndSet(false, true) && numThreads > 0) {
+
+      LOG.log(Level.FINEST, "Closing ThreadPoolStage {0}: begin", this.name);
+
       executor.shutdown();
-      if (!executor.awaitTermination(shutdownTimeout, TimeUnit.MILLISECONDS)) {
-        LOG.log(Level.WARNING, "Executor did not terminate in " + shutdownTimeout + "ms.");
+
+      boolean isTerminated = false;
+      try {
+        isTerminated = executor.awaitTermination(SHUTDOWN_TIMEOUT, TimeUnit.MILLISECONDS);
+      } catch (final InterruptedException ex) {
+        LOG.log(Level.WARNING, "Interrupted closing ThreadPoolStage " + this.name, ex);
+      }
+
+      if (!isTerminated) {
         final List<Runnable> droppedRunnables = executor.shutdownNow();
-        LOG.log(Level.WARNING, "Executor dropped " + droppedRunnables.size() + " tasks.");
+        LOG.log(Level.WARNING,
+            "Closing ThreadPoolStage {0}: Executor did not terminate in {1} ms. Dropping {2} tasks",
+            new Object[] {this.name, SHUTDOWN_TIMEOUT, droppedRunnables.size()});
       }
+
```

Comment:
No, `.shutdownNow()` cannot guarantee that all threads are down - it is possible for the user to submit a `Runnable` that will not respond to `Thread.interrupt()` and thus never terminate.

A combination of `.awaitTermination()` (that stops accepting new tasks and interrupts threads after timeout) and `.shutdownNow()` (that closes the threads and returns tasks still left in the queue after termination) is our best effort to shut down the thread pool. It must be combined with the policy that all user `Runnable` tasks shall never catch and ignore the `InterruptedException`.